### PR TITLE
Tabular: Added user-specified groups parameter, fixed HPO for RF and KNN

### PR DESCRIPTION
**Pull request by Innixma**
*Issue #, if available:*

#1174, #1120, #617, #901

*Description of changes:*

- Added user-specified groups parameter to enable custom fold-splits.
- Fixed HPO for RF and KNN (previously it would crash and not train the models if HPO is specified).

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: tabular/src/autogluon/tabular/predictor/predictor.py [hunk 1]
```diff
@@ -91,6 +91,13 @@ class TabularPredictor:
         If True, then weighted metrics will be reported based on the sample weights provided in the specified `sample_weight` (in which case `sample_weight` column must also be present in test data).
         In this case, the 'best' model used by default for prediction will also be decided based on a weighted version of evaluation metric.
         Note: we do not recommend specifying `weight_evaluation` when `sample_weight` is 'auto_weight' or 'balance_weight', instead specify appropriate `eval_metric`.
+    groups : str, default = None
+        [Experimental] If specified, AutoGluon will use the column named the value of groups in `train_data` during `.fit` as the data splitting indices for the purposes of bagging.
```
Comment (jwmueller) [inline tabular/src/autogluon/tabular/predictor/predictor.py:95]: may be better left as kwarg while it is experimental?

File: tabular/src/autogluon/tabular/predictor/predictor.py [hunk 2]
```diff
@@ -91,6 +91,13 @@ class TabularPredictor:
...
+        This column will not be used as a feature during model training.
+        The data will be split via `sklearn.model_selection.LeaveOneGroupOut`.
+        Use this option to control the exact split indices AutoGluon uses.
```
Comment (jwmueller) [inline tabular/src/autogluon/tabular/predictor/predictor.py:98]: probably need to clarify what the `groups` column values should look like. And how it works with bagging vs no-bagging

Comment (jwmueller) [reply tabular/src/autogluon/tabular/predictor/predictor.py:98]: could be helpful to provide a motivating example: "For example, if you want your data folds to preserve adjacent rows in the table (without shuffling), then the `groups` column should look like ..."

**Comment by szha** [comment]
Job PR-1224-3 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1224/3/index.html

File: tabular/src/autogluon/tabular/predictor/predictor.py (hunk 2 above)
Comment (Innixma) [reply tabular/src/autogluon/tabular/predictor/predictor.py:99]: added example and more clarifications

File: tabular/src/autogluon/tabular/predictor/predictor.py (hunk 1 above)
Comment (Innixma) [reply tabular/src/autogluon/tabular/predictor/predictor.py:95]: Since this feature is asked about so much, I'd rather it be highly visible so it isn't missed. By being experimental it does fully work without issue, it's just if the user specifies an invalid or weird grouping, edge cases can occur that may result in strange errors (hard to detect if groups are invalid or not prior to trying). This is mostly related to cases with very small amounts of data.

**Comment by szha** [comment]
Job PR-1224-4 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1224/4/index.html

**Comment by szha** [comment]
Job PR-1224-5 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1224/5/index.html

File: tabular/src/autogluon/tabular/predictor/predictor.py [hunk 3]
```diff
@@ -91,6 +91,16 @@ class TabularPredictor:
...
         Note: we do not recommend specifying `weight_evaluation` when `sample_weight` is 'auto_weight' or 'balance_weight', instead specify appropriate `eval_metric`.
+    groups : str, default = None
+        [Experimental] If specified, AutoGluon will use the column named the value of groups in `train_data` during `.fit` as the data splitting indices for the purposes of bagging.
+        This column will not be used as a feature during model training.
+        This parameter is ignored if bagging is not enabled. To instead specify a custom validation set with bagging disabled, specify `tuning_data` in `.fit`.
+        The data will be split via `sklearn.model_selection.LeaveOneGroupOut`.
+        Use this option to control the exact split indices AutoGluon uses.
+        It is not recommended to use this option unless it is required for very specific situations.
+        Bugs may arise from edge cases if the provided groups are not valid to properly train models, such as if not all classes are present during training in multiclass classification. It is up to the user to sanitize their groups.
+
```
Comment (gradientsky) [inline tabular/src/autogluon/tabular/predictor/predictor.py:102]: There are couple cases to worry here:
1) Heavy classes imbalances between groups would results with incorrect time limits estimates and potentially memory errors on 'heavy' folds. Sanity check for such imbalance might be helpful.
2) sub-folds might have just a subset of training labels and bags need to be aligned on label space if some form of label encoding is used.
```
...,0,A
...,0,B
...,0,C
...,1,C
...,1,D
...,1,D
```
First fold would know only about A,B,C; 2nd about C,D. Is there handling which would align the fold predictions into an aggregate prediciton?

File: core/src/autogluon/core/utils/utils.py [hunk 4]
```diff
@@ -40,40 +40,66 @@ def get_gpu_count():
...
+        self.groups = groups
+        if splitter_cls is None:
+            splitter_cls = self._get_splitter_cls()
+        self._splitter = self._get_splitter(splitter_cls)
+
+    def _get_splitter_cls(self):
+        if self.groups is not None:
+            num_groups = len(self.groups.unique())
+            if self.n_repeats != 1:
+                raise AssertionError(f'n_repeats must be 1 when split groups are specified. (n_repeats={self.n_repeats})')
```
Comment (gradientsky) [inline core/src/autogluon/core/utils/utils.py:64]: Does it makes sense to raise exception - we can log a warning and reset it to 1 automatically.

File: core/src/autogluon/core/utils/utils.py [hunk 5]
```diff
@@ -40,40 +40,66 @@ def get_gpu_count():
...
+            self.n_splits = num_groups
+            splitter_cls = LeaveOneGroupOut
```
Comment (gradientsky) [inline core/src/autogluon/core/utils/utils.py:66]: log a warning that the feature is experimental

File: tabular/src/autogluon/tabular/predictor/predictor.py (hunk 3 above)
Comment (Innixma) [reply tabular/src/autogluon/tabular/predictor/predictor.py:102]: 1. This is very hard to fix, even if we knew there was imbalance it wouldn't necessarily help us that much.

2. Worst case scenario which causes crash is if a class is only present in 1 fold, meaning that a model that uses that fold as validation won't see that class in training. This is what I'd consider an invalid grouping, and probably isn't worth trying to work with (unless we later determine that there are valid use-cases for this being the case)

File: core/src/autogluon/core/utils/utils.py (hunk 5 above)
Comment (Innixma) [reply core/src/autogluon/core/utils/utils.py:66]: This would log spam: This is repeated for every bag.

File: core/src/autogluon/core/utils/utils.py (hunk 4 above)
Comment (Innixma) [reply core/src/autogluon/core/utils/utils.py:64]: I prefer an exception, n_repeats is simply not valid for this situation and should never be passed in this fashion.

File: core/src/autogluon/core/models/ensemble/bagged_ensemble_model.py [hunk 6]
```diff
@@ -194,6 +198,41 @@ def _fit(self,
...
+                             groups):
+        if groups is not None:
+            if self._n_repeats_finished != 0:
+                raise AssertionError('Bagged models cannot call fit with `groups` specified when a full k-fold set has already been fit.')
+            if n_repeats > 1:
+                raise AssertionError('Cannot perform repeated bagging with `groups` specified.')
+            return
+
+        if k_fold_end is None:
+            k_fold_end = k_fold
```
[the remark this answers is not in the document]
Comment (Innixma) [reply core/src/autogluon/core/models/ensemble/bagged_ensemble_model.py:216]: This occurs elsewhere. At the point where this validate is called, groups can also be present, meaning k_fold itself can still change, so we don't want to change k_fold_end to the old k_fold value.

**Review by jwmueller — APPROVED** [review]
LGTM
